helix-view/src/clipboard.rs

```diff
@@ -1,413 +1,521 @@
+mod noop {
+    use super::*;
 
-#[cfg(not(target_os = "windows"))]
-pub mod provider {
-    use super::{ClipboardProvider, ClipboardType};
-    use anyhow::Result;
-    use std::borrow::Cow;
+    #[derive(Debug, Clone)]
+    pub enum ClipboardConfig {}
 
-    #[cfg(feature = "term")]
-    mod osc52 {
-        use {super::ClipboardType, crate::base64};
+    impl ClipboardConfig {
+        pub fn detect() -> Self {
+            Self
+        }
 
-        #[derive(Debug)]
-        pub struct SetClipboardCommand {
-            encoded_content: String,
-            clipboard_type: ClipboardType,
+        pub fn name(&self) -> Cow<str> {
+            "none".into()
         }
 
-        impl SetClipboardCommand {
-            pub fn new(content: &str, clipboard_type: ClipboardType) -> Self {
-                Self {
-                    encoded_content: base64::encode(content.as_bytes()),
-                    clipboard_type,
-                }
-            }
+        pub fn get_contents(&self, _clipboard_type: ClipboardType) -> Result<String> {
+            Err(ClipboardError::ReadingNotSupported)
         }
 
-        impl crossterm::Command for SetClipboardCommand {
-            fn write_ansi(&self, f: &mut impl std::fmt::Write) -> std::fmt::Result {
-                let kind = match &self.clipboard_type {
-                    ClipboardType::Clipboard => "c",
-                    ClipboardType::Selection => "p",
-                };
-                // Send an OSC 52 set command: https://terminalguide.namepad.de/seq/osc-52/
-                write!(f, "\x1b]52;{};{}\x1b\\", kind, &self.encoded_content)
-            }
+        pub fn set_contents(&self, _content: &str, _clipboard_type: ClipboardType) -> Result<()> {
+            Ok(())
         }
     }
+}
 
-    #[derive(Debug)]
-    pub struct FallbackProvider {
-        buf: String,
-        primary_buf: String,
-    }
+#[cfg(not(target_arch = "wasm32"))]
+mod external {
+    use super::*;
 
-    impl FallbackProvider {
-        pub fn new() -> Self {
-            #[cfg(feature = "term")]
-            log::debug!(
-                "No native clipboard provider found. Yanking by OSC 52 and pasting will be internal to Helix"
-            );
-            #[cfg(not(feature = "term"))]
-            log::warn!(
-                "No native clipboard provider found! Yanking and pasting will be internal to Helix"
-            );
-            Self {
-                buf: String::new(),
-                primary_buf: String::new(),
-            }
-        }
+    #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+    pub struct Command {
+        command: String,
+        #[serde(default = "Vec::new")]
+        args: Vec<String>,
     }
 
-    impl Default for FallbackProvider {
-        fn default() -> Self {
-            Self::new()
-        }
+    #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+    #[serde(rename_all = "kebab-case")]
+    pub struct CommandProvider {
```

Comment:
As you say the StaticCommandProvider + CommandProvider part is not ideal - we can combine these types. You can use a `Cow<'lifetime, BorrowedT>` to represent a type that is either owned or borrowed, so `Command` and `CommandProvider` can be written as

```rust
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Command<'a> {
    command: Cow<'a, str>,
    #[serde(default)]
    args: Cow<'a, [Cow<'a, str>]>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct CommandProvider<'a> {
    yank: Command<'a>,
    paste: Command<'a>,
    yank_primary: Option<Command<'a>>,
    paste_primary: Option<Command<'a>>,
}
```

For commands that we know at compile time the type will end up with strs and slices borrowed for `'static` and then for custom commands that we need to allocate at runtime we'll use the `Cow::Owned` branch of the enum. For the current `ClipboardConfig` enum this branch would then be `Custom(Box<CommandProvider<'static>>)` (Box because the CommandProvider struct is much larger than the other fields of `ClipboardConfig` - this is a clippy lint).

Then functions like `yank_from_builtin` can take a `provider: CommandProvider<'static>` parameter and `execute_command` can take `cmd: &Command<'_>`. That's the nice part of using a Cow for this type - we can support passing static commands and dynamically created ones with the same signature

Reply:
Actually it is probably easier to work with these types if we explicitly use the `'static` lifetime:

```rust
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Command {
    command: Cow<'static, str>,
    #[serde(default)]
    args: Cow<'static, [Cow<'static, str>]>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct CommandProvider {
    yank: Command,
    paste: Command,
    yank_primary: Option<Command>,
    paste_primary: Option<Command>,
}
```

then we can drop the lifetime bounds part anywhere we mention this type, so functions would take `&Command` or `CommandProvider`